### PR TITLE
restructure sns and dynamo clients

### PR DESCRIPTION
**Pull request by MancunianSam**
I've changed these around so they have a trait and the apply method
creates an implementation of that trait. This will make it easier to
provide our own test traits in the unit tests.

I've also got rid of all the <: Product

### CODE REVIEW

**Review by techncl — DISMISSED** [review]
Looks good! Just a few comments

File: dynamodb/src/main/scala/uk/gov/nationalarchives/DADynamoDBClient.scala [hunk 1]
```diff
@@ -95,17 +69,31 @@ class DADynamoDBClient[F[_]: Async](dynamoDBClient: DynamoDbAsyncClient):
...
+    *   This is not passed in directly. You construct either a Query[_] or a ConditionExpression instance. This is then
+    *   converted to RequestCondition by the implicits in the companion object.
+    * @param returnTypeFormat
+    *   An implicit scanamo formatter for return type U
+    * @tparam U
+    *   The return type for the function
+    * @return
+    *   A list of type U wrapped in the F effect
+    */
+  def queryItems[U](tableName: String, gsiName: String, requestCondition: RequestCondition)(using
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply dynamodb/src/main/scala/uk/gov/nationalarchives/DADynamoDBClient.scala:90]: Yeah ok, that's a bit tidier. I've changed it.

File: dynamodb/src/main/scala/uk/gov/nationalarchives/DADynamoDBClient.scala [hunk 2]
```diff
@@ -22,16 +22,13 @@ import scala.language.{implicitConversions, postfixOps}
...
   *   Type of the effect
   */
-class DADynamoDBClient[F[_]: Async](dynamoDBClient: DynamoDbAsyncClient):
-  extension [T](completableFuture: CompletableFuture[T])
-    private def liftF: F[T] = Async[F].fromCompletableFuture(Async[F].pure(completableFuture))
+trait DADynamoDBClient[F[_]: Async]:
 
-  /** Takes a list of primary key attributes of type T and deletes the corresponding items from Dynamo
+  /** Takes a list of primary key attributes of type T and deletes the corresponding items from Dynamo. Unprocessed
+    * items will be retried
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply dynamodb/src/main/scala/uk/gov/nationalarchives/DADynamoDBClient.scala:31]: That was scalafmt. The line went over the max 120 characters. I'll leave it as it matches everything else.

File: dynamodb/src/main/scala/uk/gov/nationalarchives/DADynamoDBClient.scala [hunk 3]
```diff
@@ -279,4 +150,170 @@ object DADynamoDBClient:
       conditionalExpression: Option[String] = None
   )
 
-  def apply[F[_]: Async]() = new DADynamoDBClient[F](dynamoDBClient)
+  def apply[F[_]: Async](): DADynamoDBClient[F] = {
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply dynamodb/src/main/scala/uk/gov/nationalarchives/DADynamoDBClient.scala:153]: Yeah ok. I've changed this one and the one for the SNS client.

File: dynamodb/src/main/scala/uk/gov/nationalarchives/DADynamoDBClient.scala [hunk 4]
```diff
@@ -76,7 +75,7 @@ trait DADynamoDBClient[F[_]: Async]:
   /** @param tableName
     *   The name of the table
     * @param gsiName
```
Comment (techncl) [inline dynamodb/src/main/scala/uk/gov/nationalarchives/DADynamoDBClient.scala:77]: This name doesn't match the parameter name

Comment (MancunianSam) [reply dynamodb/src/main/scala/uk/gov/nationalarchives/DADynamoDBClient.scala]: Ah yeah, good spot, I've changed it.

**Pull request merged**